ccl/changefeedccl: TestChangefeedOnlyInitialScanCSV failed

ccl/changefeedccl.TestChangefeedOnlyInitialScanCSV [failed](https://mesolite.cluster.engflow.com/invocations/default/24909baa-f0e1-4c87-8dab-35b58c1550ce?testReportRun=3&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on release-25.4.0-rc @ [4f4179243a351bc48c00dd9243dc56b8af98e210](https://github.com/cockroachdb/cockroach/commits/4f4179243a351bc48c00dd9243dc56b8af98e210):


```
=== RUN   TestChangefeedOnlyInitialScanCSV
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestChangefeedOnlyInitialScanCSV3971836438
    test_log_scope.go:82: use -show-logs to present logs inline
    changefeed_test.go:10062: -- test log scope end --
test logs left over in: outputs.zip/logTestChangefeedOnlyInitialScanCSV3971836438
--- FAIL: TestChangefeedOnlyInitialScanCSV (2.18s)
=== RUN   TestChangefeedOnlyInitialScanCSV/kafka/initial_backfill_only_with_csv_multiple_tables
    changefeed_test.go:10044: 
        	Error Trace:	pkg/ccl/changefeedccl/changefeed_test.go:10044
        	            				pkg/ccl/changefeedccl/changefeed_test.go:10056
        	Error:      	Not equal: 
        	            	expected: 6
        	            	actual  : 2
        	Test:       	TestChangefeedOnlyInitialScanCSV/kafka/initial_backfill_only_with_csv_multiple_tables
--- FAIL: TestChangefeedOnlyInitialScanCSV/kafka/initial_backfill_only_with_csv_multiple_tables (0.45s)
=== RUN   TestChangefeedOnlyInitialScanCSV/kafka
    helpers_test.go:1394: making server as system tenant
    helpers_test.go:1470: making kafka feed factory
    testfeed_test.go:341: creating external connection
    testfeed_test.go:344: ran create external connection
    testfeed_test.go:341: creating external connection
    testfeed_test.go:344: ran create external connection
    testfeed_test.go:341: creating external connection
    testfeed_test.go:344: ran create external connection
--- FAIL: TestChangefeedOnlyInitialScanCSV/kafka (1.77s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=3</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedOnlyInitialScanCSV.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-56392

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -10132,6 +10132,8 @@ func TestChangefeedOnlyInitialScanCSV(t *testing.T) {
 			t.Run(testName, func(t *testing.T) {
 				sqlDB.Exec(t, "CREATE TABLE foo (id INT PRIMARY KEY, name STRING)")
 				sqlDB.Exec(t, "CREATE TABLE bar (id INT PRIMARY KEY, name STRING)")
+				defer sqlDB.Exec(t, "DROP TABLE foo")
+				defer sqlDB.Exec(t, "DROP TABLE bar")
 
 				sqlDB.Exec(t, "INSERT INTO foo VALUES (1, 'Alice'), (2, 'Bob'), (3, 'Carol')")
 				sqlDB.Exec(t, "INSERT INTO bar VALUES (1, 'a'), (2, 'b'), (3, 'c')")
@@ -10139,36 +10141,26 @@ func TestChangefeedOnlyInitialScanCSV(t *testing.T) {
 				sqlDB.CheckQueryResultsRetry(t, `SELECT count(*) FROM foo,bar`, [][]string{{`9`}})
 
 				feed := feed(t, f, testData.changefeedStmt)
+				defer closeFeed(t, feed)
 
 				sqlDB.Exec(t, "INSERT INTO foo VALUES (4, 'Doug'), (5, 'Elaine'), (6, 'Fred')")
 				sqlDB.Exec(t, "INSERT INTO bar VALUES (4, 'd'), (5, 'e'), (6, 'f')")
 
-				var actualMessages []string
-				g := ctxgroup.WithContext(context.Background())
-				g.Go(func() error {
-					for {
-						m, err := feed.Next()
-						if err != nil {
-							return err
-						}
-						if len(m.Resolved) > 0 {
-							continue
-						}
-						actualMessages = append(actualMessages, string(m.Value))
+				err := withTimeout(feed, assertPayloadsTimeout(), func(ctx context.Context) error {
+					msgs, err := readNextMessages(ctx, feed, len(testData.expectedPayload))
+					if err != nil {
+						return err
 					}
+					actual := make([]string, 0, len(testData.expectedPayload))
+					for _, m := range msgs {
+						actual = append(actual, string(m.Value))
+					}
+					sort.Strings(testData.expectedPayload)
+					sort.Strings(actual)
+					require.Equal(t, testData.expectedPayload, actual)
+					return nil
 				})
-				defer func(expectedPayload []string) {
-					closeFeed(t, feed)
-					sqlDB.Exec(t, `DROP TABLE foo`)
-					sqlDB.Exec(t, `DROP TABLE bar`)
-					_ = g.Wait()
-					require.Equal(t, len(expectedPayload), len(actualMessages))
-					sort.Strings(expectedPayload)
-					sort.Strings(actualMessages)
-					for i := range expectedPayload {
-						require.Equal(t, expectedPayload[i], actualMessages[i])
-					}
-				}(testData.expectedPayload)
+				require.NoError(t, err)
 
 				jobFeed := feed.(cdctest.EnterpriseTestFeed)
 				require.NoError(t, jobFeed.WaitForState(func(s jobs.State) bool {

--- a/pkg/ccl/changefeedccl/helpers_test.go
+++ b/pkg/ccl/changefeedccl/helpers_test.go
@@ -115,7 +115,7 @@ func readNextMessages(
 	lastMessage := timeutil.Now()
 	for len(actual) < numMessages {
 		if ctx.Err() != nil {
-			return nil, ctx.Err()
+			return nil, errors.Wrapf(ctx.Err(), "got %d of %d expected changefeed messages", len(actual), numMessages)
 		}
 		if log.V(1) {
 			log.Changefeed.Infof(context.Background(), "about to read a message (%d out of %d)", len(actual), numMessages)
@@ -131,7 +131,7 @@ func readNextMessages(
 		}
 		lastMessage = timeutil.Now()
 		if err != nil {
-			return nil, err
+			return nil, errors.Wrapf(err, "got %d of %d expected changefeed messages", len(actual), numMessages)
 		}
 		if m == nil {
 			return nil, errors.AssertionFailedf(`expected message`)

